### DISCUSSION TITLE
2023/04/the-best-way-to-get-nvm-working-in-ci/cd-systems/

### DISCUSSION THREAD

**Discussion by giscus** [tgagor/tgagor.github.io#3] [Announcements]
# 2023/04/the-best-way-to-get-nvm-working-in-ci/cd-systems/

TL;DR While reasoning is important, readers may not be interested in all the frustrations I experienced while figuring out how to get things done. If you’re lo

https://gagor.pl/2023/04/the-best-way-to-get-nvm-working-in-ci/cd-systems/

**Comment by kraftman** [comment]
It's hard to tell exactly what you're doing with Jenkins, but generally people don't deploy multiple projects to the same docker container, each project has it's own docker file, which specifies it's own node version. Your pipeline would build and rerun the docker file to test, and then that gets deployed to production. This keeps your testing and production environments as similar as possible to avoid surprises.

**Reply by tgagor** [reply]
That's true, it's hard to explain without full context. We're not installing multiple apps in the single container. We just have a lot of apps with a variety of requirements and what we tried to resolve here is: how to provide support for different versions of Node?
NVM makes it simple on local machine, but because it's designed for interactive shells and it's not obvious how to use it in CI/CD.

Running final Dockerfile is more like a smoke test, but we often use unit tests or e2e tests too. Those require additional dependencies and you don't want them in your final Docker image. That's why we keep our Jenkins agents pretty similar to the base images that are used for apps.

I hope it's more clear now.
